packages/dds/tree/docs/wip/Container and Application Schema.md

```diff
@@ -0,0 +1,194 @@
+
+-   pure schema-on-read: no customization of container schema.
+-   almost pure schema-on-write: only using schema-on-read to assist with schema migration so their application schema doesn't have to deal with old formats (which unavoidably pile up in the container schema)
+    They may additionally use schema-on-read to enforce/fix some invariants the container schema system isn't expressive enough to declare (referential integrity for graph like references, numeric ranges, and any other appellation level invariants that are desired)
+
+# Example/Proposed Design
+
+This design shows an easy to implement MVP, and is not intended to have document or API compatibility with what we do longer term.
+It minimally implements the requirements from "What options to support" while being extensible in the future.
+It can be subseted for use in earlier milestones, and could be extended for later ones, see "Schedule" below.
+
+This design is intentionally minimal while still being useful.
+Features not needed for minimal strongly typed application use or the JSON and XML domain schema are not included.
+
+## Schema Representation
+
+Include a simple non-contextual schema system, usable as a MVP for both `container schema` and `application schema`.
+
+We can start with:
+
+```typescript
+enum Multiplicity {
+    Value,
+    Optional,
+    Sequence,
+}
+
+enum Value {
+    Nothing,
+    Number,
+    String,
+    Boolean,
+    Serializable,
+}
+
+interface FieldContent {
+    multiplicity: Multiplicity;
+    types?: Type[];
+}
+
+interface Field {
+    content: FieldContent;
+    name: string;
+}
+
+interface Type {
+    name: string;
+    fields: Field[];
+    extraFields: FieldContent;
+    value: Value;
+}
+```
+
+These `Types` can be added as `container schema` as part of an edit op, which is considered conflicted if it tries to add a type that has a conflicting `name` and is not equal to the existing one.
+
+This intentionally does not support any concept of versioning or name-spacing: users of it can include versions and namespaces in the `name` if they want, or features for them could be added later.
+
+While a schema must be added to the document as a container schema to use the type
+(otherwise adding a new type could break existing data which might not even be downloaded on the current client),
+it's possible to add a schema that is compatible with all possible data.
+Applications which wish to rely entirely on schema-on-read for some or all of their data can use this pattern for all `container schema` and only use their actual developer authored schema as `application schema`:
+
+```typescript
+const anyField: FieldContent = { multiplicity: Multiplicity.sequence };
+function makeAnyType(name: string): Type {
+    return { name, extraFields: anyField };
+}
+```
+
+## Use as `Container Schema`
+
+-   Document load: The application can check that their application schema match container schema, and check that the root has supported content.
```

Comment:
I think the answer is "yes", but your terminology is a bit unclear:

There is a variance issue here: since we want a read/write APIs, the relation here is invariant: if the application schema is exactly the same as the container schema, then the app should be happy using the document directly (thats the case being addresses here) and it does not matter which schema is used. If the application's schema is more permissive, then there may be issues when editing, and if it is more restrictive, there may be issues when reading.

Thus I think "conforms to" should only be used to relate data to schema, not schema to schema. For schema, we should say if we mean equal to, or if one is more constrained (permits a subset of the documents) or less constrained (permits a superset of the documents)

Reply:
I have updated the document to refer to the example code which provides an implementation of this. Hopefully that can clear it up.